File: backend/src/ledger_sync/api/analytics.py
"""Analytics API endpoints for insights and statistics."""

from typing import Annotated, Any

from fastapi import APIRouter, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from ledger_sync.api.analytics_helpers import (
    _get_sql_account_totals,
    _get_sql_category_totals,
    _get_sql_monthly_data,
    _get_sql_totals,
    get_filtered_transactions,
)
from ledger_sync.api.deps import CurrentUser, DatabaseSession
from ledger_sync.core import calculator
from ledger_sync.core.time_filter import TimeRange
from ledger_sync.db.models import TransactionType, User, UserPreferences

router = APIRouter(prefix="/api/analytics", tags=["analytics"])

TIME_RANGE_FILTER_DESC = "Time range filter"


def _load_preferences(db: Session, user: User) -> UserPreferences | None:
    """Read a user's preferences row, or ``None`` when they have none yet.

    Read-only counterpart to ``preferences_helpers._get_or_create_preferences``:
    a GET must not write a defaults row as a side effect. Consumers treat
    ``None`` as "use the shipped defaults", so a missing row is not an error.
    """
    stmt = select(UserPreferences).where(UserPreferences.user_id == user.id).limit(1)
    return db.execute(stmt).scalar_one_or_none()


@router.get("/overview")
def get_overview(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get overview statistics: income, expenses, net change, best/worst month.

    ``capital_losses`` is published as its own figure and is already netted out
    of ``net_change`` by ``_get_sql_totals``, so this response reconciles with
    ``/api/calculations/totals`` (``net_change == net_savings``) instead of
    overstating the net position by the loss amount.
    """
    totals = _get_sql_totals(db, current_user, time_range)

    if totals["transaction_count"] == 0:
        return {
            "total_income": 0,
            "total_expenses": 0,
            "capital_losses": 0,
            "net_change": 0,
            "best_month": None,
            "worst_month": None,
            "asset_allocation": [],
            "transaction_count": 0,
        }

    # SQL-based aggregations
    monthly_data = _get_sql_monthly_data(db, current_user, time_range)
    best_worst = calculator.find_best_worst_months(monthly_data)
    account_activity = _get_sql_account_totals(db, current_user, time_range)

    # Format asset allocation
    asset_allocation: list[dict[str, Any]] = [
        {"account": account, "balance": activity} for account, activity in account_activity.items()
    ]
    asset_allocation.sort(key=lambda x: float(x["balance"]), reverse=True)

    return {
        "total_income": totals["total_income"],
        "total_expenses": totals["total_expenses"],
        "capital_losses": totals["capital_losses"],
        "net_change": totals["net_change"],
        "best_month": best_worst["best_month"],
        "worst_month": best_worst["worst_month"],
        "asset_allocation": asset_allocation,
        "transaction_count": totals["transaction_count"],
    }


@router.get("/behavior")
def get_behavior(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get spending behavior metrics."""
    transactions = get_filtered_transactions(db, current_user, time_range)

    if not transactions:
        return {
            "avg_transaction_size": 0,
            "spending_frequency": 0,
            "convenience_spending_pct": 0,
            "lifestyle_inflation": 0,
            "top_categories": [],
        }

    # Use calculator for metrics
    lifestyle_inf = calculator.calculate_lifestyle_inflation(transactions)
    convenience_data = calculator.calculate_convenience_spending(transactions)
    convenience_pct = convenience_data["convenience_pct"]
    category_totals = calculator.group_by_category(transactions)

    # Calculate average transaction size and frequency (specific to this endpoint)
    expenses = [t for t in transactions if t.type == TransactionType.EXPENSE]
    if not expenses:
        return {
            "avg_transaction_size": 0,
            "spending_frequency": 0,
            "convenience_spending_pct": convenience_pct,
            "lifestyle_inflation": lifestyle_inf,
            "top_categories": [],
        }

    avg_transaction_size = sum(float(t.amount) for t in expenses) / len(expenses)

    # Spending frequency (transactions per month)
    dates = [t.date for t in expenses]
    min_date = min(dates)
    max_date = max(dates)
    months_span = (max_date.year - min_date.year) * 12 + (max_date.month - min_date.month) + 1
    spending_frequency = len(expenses) / months_span if months_span > 0 else 0

    # Top spending categories
    top_categories: list[dict[str, Any]] = [
        {"category": cat, "amount": amt}
        for cat, amt in category_totals.items()
        if cat  # Only include expense categories
    ]
    top_categories.sort(key=lambda x: float(x["amount"]), reverse=True)

    return {
        "avg_transaction_size": avg_transaction_size,
        "spending_frequency": spending_frequency,
        "convenience_spending_pct": convenience_pct,
        "lifestyle_inflation": lifestyle_inf,
        "top_categories": top_categories[:10],
    }


@router.get("/trends")
def get_trends(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get spending and income trends over time.

    ``consistency_score`` carries ``consistency_measurable``: the score function
    returns a flat 100.0 when a coefficient of variation is undefined (one month
    of history, or a zero mean), and 100 reads as the BEST possible result. A
    client that renders the number without checking the flag tells a brand-new
    user their spending is perfectly consistent.
    """
    transactions = get_filtered_transactions(db, current_user, time_range)

    if not transactions:
        return {
            "monthly_trends": [],
            "surplus_trend": [],
            "consistency_score": 0,
            "consistency_measurable": False,
        }

    # Use calculator for metrics
    monthly_data = calculator.group_by_month(transactions)
    monthly_expenses = [data["expenses"] for data in monthly_data.values()]
    consistency_score = calculator.calculate_consistency_score(monthly_expenses)
    consistency_measurable = calculator.is_measurable_consistency(monthly_expenses)

    # Format monthly trends
    monthly_trends = [
        {
            "month": month,
            "income": data["income"],
            "expenses": data["expenses"],
            "surplus": data["income"] - data["expenses"],
        }
        for month, data in sorted(monthly_data.items())
    ]

    # Surplus trend for easy charting
    surplus_trend = [
        {
            "month": trend["month"],
            "surplus": trend["surplus"],
        }
        for trend in monthly_trends
    ]

    return {
        "monthly_trends": monthly_trends,
        "surplus_trend": surplus_trend,
        "consistency_score": consistency_score,
        "consistency_measurable": consistency_measurable,
    }


@router.get("/wrapped")
def get_yearly_wrapped(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get yearly wrapped insights - text-based narratives."""
    transactions = get_filtered_transactions(db, current_user, time_range)

    if not transactions:
        return {"insights": []}

    # Use calculator for metrics
    totals = calculator.calculate_totals(transactions)
    monthly_data = calculator.group_by_month(transactions)
    best_worst = calculator.find_best_worst_months(monthly_data)
    savings_rate = calculator.calculate_savings_rate(
        totals["total_income"],
        totals["total_expenses"],
    )
    daily_rate = calculator.calculate_daily_spending_rate(transactions)

    expenses = [t for t in transactions if t.type == TransactionType.EXPENSE]
    income_txns = [t for t in transactions if t.type == TransactionType.INCOME]

    insights = []

    # Total spending insight
    insights.append(
        {
            "title": "Total Spending",
            "value": f"₹{totals['total_expenses']:,.2f}",
            "description": (
                f"You spent ₹{totals['total_expenses']:,.2f} across {len(expenses)} transactions"
            ),
        },
    )

    # Total income insight
    insights.append(
        {
            "title": "Total Income",
            "value": f"₹{totals['total_income']:,.2f}",
            "description": (
                f"You earned ₹{totals['total_income']:,.2f} from {len(income_txns)} sources"
            ),
        },
    )

    # Biggest expense
    if expenses:
        biggest = max(expenses, key=lambda t: float(t.amount))
        insights.append(
            {
                "title": "Biggest Expense",
                "value": f"₹{float(biggest.amount):,.2f}",
                "description": (
                    f"Your largest expense was ₹{float(biggest.amount):,.2f} in {biggest.category}"
                ),
            },
        )

    # Most frequent category
    if expenses:
        category_counts: dict[str, int] = {}
        for t in expenses:
            category_counts[t.category] = category_counts.get(t.category, 0) + 1
        most_frequent = max(category_counts.items(), key=lambda x: x[1])
        insights.append(
            {
                "title": "Most Frequent Category",
                "value": most_frequent[0],
                "description": (f"You made {most_frequent[1]} transactions in {most_frequent[0]}"),
            },
        )

    # Best month
    if best_worst["best_month"]:
        best = best_worst["best_month"]
        insights.append(
            {
                "title": "Best Month",
                "value": best["month"],
                "description": (
                    f"Your best month was {best['month']} with a surplus of ₹{best['surplus']:,.2f}"
                ),
            },
        )

    # Savings rate
    insights.append(
        {
            "title": "Savings Rate",
            "value": f"{savings_rate:.1f}%",
            "description": f"You saved {savings_rate:.1f}% of your income",
        },
    )

    # Daily average spending
    insights.append(
        {
            "title": "Daily Average",
            "value": f"₹{daily_rate:,.2f}",
            "description": f"You spent an average of ₹{daily_rate:,.2f} per day",
        },
    )

    return {"insights": insights}


# New Enhanced Endpoints for Phase 2 Expansion


@router.get("/kpis")
def get_kpis(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get all KPI metrics in one call.

    Two of these numbers have an undefined case that returns a plausible-looking
    value rather than an error, so each ships with a companion flag:

    * ``consistency_measurable`` -- ``consistency_score`` is a flat 100.0 (the
      BEST possible reading) whenever a coefficient of variation is undefined.
    * ``velocity_comparable`` -- ``spending_velocity`` is 0.0 when there is no
      history outside the recent window, which reads as "spending is 100% down".

    A client that renders either number without its flag reports a conclusion the
    data does not support.
    """
    transactions = get_filtered_transactions(db, current_user, time_range)

    if not transactions:
        return {
            "savings_rate": 0,
            "daily_spending_rate": 0,
            "monthly_burn_rate": 0,
            "spending_velocity": 0,
            "velocity_comparable": False,
            "category_concentration": 0,
            "consistency_score": 0,
            "consistency_measurable": False,
            "lifestyle_inflation": 0,
            "convenience_spending_pct": 0,
        }

    totals = calculator.calculate_totals(transactions)
    monthly_data = calculator.group_by_month(transactions)
    monthly_expenses = [data["expenses"] for data in monthly_data.values()]
    category_totals = calculator.group_by_category(transactions)
    spending_velocity = calculator.calculate_spending_velocity(transactions)
    convenience_data = calculator.calculate_convenience_spending(transactions)

    return {
        "savings_rate": calculator.calculate_savings_rate(
            totals["total_income"],
            totals["total_expenses"],
        ),
        "daily_spending_rate": calculator.calculate_daily_spending_rate(transactions),
        "monthly_burn_rate": calculator.calculate_monthly_burn_rate(transactions),
        "spending_velocity": spending_velocity["velocity_ratio"]
        * 100,  # Convert ratio to percentage
        "velocity_comparable": spending_velocity["historical_daily"] > 0,
        "category_concentration": calculator.calculate_category_concentration(
            category_totals,
        ),
        "consistency_score": calculator.calculate_consistency_score(monthly_expenses),
        "consistency_measurable": calculator.is_measurable_consistency(monthly_expenses),
        "lifestyle_inflation": calculator.calculate_lifestyle_inflation(transactions),
        "convenience_spending_pct": convenience_data["convenience_pct"],
    }


@router.get("/charts/income-expense")
def get_income_expense_chart(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get data for income vs expense doughnut chart.

    A classified realised loss gets its own slice. Dropping it from "Expenses"
    without adding it back left the chart short by the loss amount, so the two
    slices no longer accounted for where the money went. The slice is omitted
    entirely when nothing is classified, which is every user's default state.
    """
    totals = _get_sql_totals(db, current_user, time_range)

    data: list[dict[str, Any]] = [
        {"name": "Income", "value": totals["total_income"]},
        {"name": "Expenses", "value": totals["total_expenses"]},
    ]
    if totals["capital_losses"]:
        data.append({"name": "Capital Losses", "value": totals["capital_losses"]})

    return {"data": data}


@router.get("/charts/categories")
def get_categories_chart(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
    limit: Annotated[int, Query(description="Number of top categories to return")] = 10,
) -> dict[str, Any]:
    """Get data for top categories bar chart."""
    category_totals = _get_sql_category_totals(db, current_user, time_range)

    # Sort and limit
    sorted_categories = sorted(category_totals.items(), key=lambda x: x[1], reverse=True)[:limit]

    return {"data": [{"category": cat, "amount": amt} for cat, amt in sorted_categories]}


@router.get("/charts/monthly-trends")
def get_monthly_trends_chart(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get data for monthly trends line chart.

    ``net`` subtracts ``capital_losses`` as well as ``expenses``: the loss is not
    consumption but the cash did leave, so a per-month ``income - expenses`` net
    would sit above the user's real balance change by exactly the loss.
    """
    monthly_data = _get_sql_monthly_data(db, current_user, time_range)

    # Format for line chart
    chart_data = [
        {
            "month": month,
            "income": data["income"],
            "expenses": data["expenses"],
            "capital_losses": data["capital_losses"],
            "net": data["income"] - data["expenses"] - data["capital_losses"],
        }
        for month, data in sorted(monthly_data.items())
    ]

    return {"data": chart_data}


@router.get("/charts/account-distribution")
def get_account_distribution_chart(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get data for account distribution doughnut chart."""
    account_totals = _get_sql_account_totals(db, current_user, time_range)

    # Sort by value
    sorted_accounts = sorted(account_totals.items(), key=lambda x: x[1], reverse=True)

    return {"data": [{"account": acc, "value": amt} for acc, amt in sorted_accounts]}


@router.get("/insights/generated")
def get_generated_insights(
    current_user: CurrentUser,
    db: DatabaseSession,
    time_range: Annotated[
        TimeRange, Query(description=TIME_RANGE_FILTER_DESC)
    ] = TimeRange.ALL_TIME,
) -> dict[str, Any]:
    """Get AI-generated insights from transaction data.

    The user's preferences row is loaded so every amount in every insight string
    carries THEIR display symbol. Constructing the engine with no argument
    silently resolved to the shipped default, so a user on USD read
    rupee-prefixed figures. The row is read, never created: this is a GET.
    """
    from ledger_sync.core.insights import InsightEngine

    transactions = get_filtered_transactions(db, current_user, time_range)

    if not transactions:
        return {"insights": []}

    engine = InsightEngine(_load_preferences(db, current_user))
    insights = engine.generate_all_insights(transactions)

    return {"insights": insights}

File: frontend/src/services/api/analytics.ts
import type { TimeRange, Transaction } from '@/types'
import { apiClient } from './client'

export interface OverviewData {
  total_income: number
  total_expenses: number
  net_change: number
  best_month: { month: string; surplus: number } | null
  worst_month: { month: string; surplus: number } | null
  asset_allocation: Array<{ account: string; balance: number }>
  transaction_count: number
}

export interface BehaviorData {
  avg_transaction_size: number
  spending_frequency: number
  convenience_spending_pct: number
  lifestyle_inflation: number
  top_categories: Array<{ category: string; amount: number }>
}

export interface TrendsData {
  monthly_trends: Array<{
    month: string
    income: number
    expenses: number
    surplus: number
  }>
  surplus_trend: Array<{ month: string; surplus: number }>
  consistency_score: number
  /**
   * False when `consistency_score` is the backend's undefined-case sentinel
   * rather than a measurement. A coefficient of variation needs two months and a
   * non-zero mean; outside that the score is a flat 100, which reads as the BEST
   * possible result. Never render the score without checking this.
   */
  consistency_measurable: boolean
}

export interface KPIData {
  savings_rate: number
  daily_spending_rate: number
  monthly_burn_rate: number
  spending_velocity: number
  /**
   * False when there is no history outside the recent window, where
   * `spending_velocity` comes back 0 -- indistinguishable from "spending is 100%
   * down" for a user whose whole ledger is inside that window.
   */
  velocity_comparable: boolean
  category_concentration: number
  consistency_score: number
  /** See `TrendsData.consistency_measurable`. */
  consistency_measurable: boolean
  lifestyle_inflation: number
  convenience_spending_pct: number
}

/**
 * Every method here sends only parameters its endpoint actually declares.
 *
 * FastAPI silently DROPS an unknown query param -- no 422, no warning -- so a
 * call that sends one compiles, type-checks, returns 200, and quietly ignores
 * the thing the caller asked for. `__tests__/analyticsParamContract.test.ts`
 * pins each method's params against the backend signature; read that file's
 * `DECLARED_PARAMS` before adding a param here.
 */
export const analyticsService = {
  /**
   * `time_range`, not a date window: `get_kpis`
   * (`backend/src/ledger_sync/api/analytics.py`) declares `time_range` and
   * nothing else. This used to take `{ start_date, end_date }`, which FastAPI
   * dropped on the floor -- so a date-filtered KPI request silently answered
   * with all-time figures. Narrowing the type here is the frontend half; giving
   * the endpoint a real date window would be a backend change.
   */
  getKPIs: async (params?: { time_range?: TimeRange }) => {
    const response = await apiClient.get<KPIData>('/api/analytics/kpis', { params })
    return response.data
  },

  /**
   * Newest-first ordering comes from the ENDPOINT, not from a param:
   * `get_transactions` hardcodes `order_by(Transaction.date.desc())` at
   * `backend/src/ledger_sync/api/transactions.py:359` and declares only
   * `start_date` / `end_date` / `limit` / `offset` (lines 332-335).
   *
   * `sort` / `sort_order` were sent here for exactly that reason and read like a
   * working sort contract, but FastAPI dropped both -- and `sort_order` is not
   * even the right name: the one endpoint that does sort,
   * `GET /api/transactions/search`, spells it `sort_by` / `sort_order`
   * (transactions.py:510-515). Do not re-add them; change the endpoint instead.
   */
  getRecentTransactions: async (limit: number = 5): Promise<Transaction[]> => {
    const response = await apiClient.get<{ data: Transaction[] } | Transaction[]>('/api/transactions', {
      params: { limit },
    })
    // Handle both old array format and new paginated format
    const data = response.data
    if (Array.isArray(data)) {
      return data.slice(0, limit)
    }
    return (data?.data || []).slice(0, limit)
  },

  getOverview: async (timeRange: TimeRange = 'all_time') => {
    const response = await apiClient.get<OverviewData>('/api/analytics/overview', {
      params: { time_range: timeRange },
    })
    return response.data
  },

  getBehavior: async (timeRange: TimeRange = 'all_time') => {
    const response = await apiClient.get<BehaviorData>('/api/analytics/behavior', {
      params: { time_range: timeRange },
    })
    return response.data
  },

  getTrends: async (timeRange: TimeRange = 'all_time') => {
    const response = await apiClient.get<TrendsData>('/api/analytics/trends', {
      params: { time_range: timeRange },
    })
    return response.data
  },
}

/**
 * Live backend endpoints this client deliberately no longer wraps.
 *
 * `GET /api/analytics/wrapped`, `/charts/income-expense`, `/charts/categories`,
 * `/charts/monthly-trends`, `/charts/account-distribution` and
 * `/insights/generated` all still exist and are still tested server-side. Their
 * client methods had zero call sites anywhere under `src/` -- including tests --
 * and three of them had a TanStack hook that nothing mounted either. Untyped
 * dead wrappers are how the `GeneratedInsight` response shape drifted wrong on
 * every single field without a compile error: nothing consumed it, so nothing
 * disagreed.
 *
 * Unlike `useKPIs` / `useOverview` / `useBehavior` -- which
 * `lib/demo/__tests__/seedReaderContract.test.tsx` pins as
 * consumerless-but-serviceable because `DEMO_ROUTES` still answers them -- none
 * of the six above had a `DEMO_ROUTES` entry, so in demo mode they resolved the
 * `[]` catch-all and would have rendered an empty chart rather than data. They
 * were not serviceable, only present.
 *
 * Re-add one as a four-line method when a page actually needs it, and add its
 * params to `DECLARED_PARAMS` in `__tests__/analyticsParamContract.test.ts`.
 */
